packages/entity/src/EntityQueryContext.ts

```diff
@@ -34,23 +36,25 @@ export class EntityNonTransactionalQueryContext extends EntityQueryContext {
   }
 
   async runInTransactionIfNotInTransactionAsync<T>(
-    transactionScope: (queryContext: EntityQueryContext) => Promise<T>
+    transactionScope: (queryContext: EntityTransactionalQueryContext) => Promise<T>
   ): Promise<T> {
-    if (this.isInTransaction()) {
-      return await transactionScope(this);
-    } else {
-      return await this.entityQueryContextProvider.runInTransactionAsync(transactionScope);
-    }
+    return await this.entityQueryContextProvider.runInTransactionAsync(transactionScope);
   }
 }
 
 export class EntityTransactionalQueryContext extends EntityQueryContext {
+  public readonly postCommitCallbacks: PostCommitCallback[] = [];
```

Comment:
API design discussion: once a post-commit callback is run, I don't think we ever want to run it again unless we guarantee the callbacks are idempotent. So I wonder if it makes sense to design this post-commit callback API in a way where running the callbacks clears them out so they are less likely to be run twice. Ex:

```ts
getPostCommitCallbacks(): PostCommitCallback[] {
  const callbacks = this.postCommitCallbacks;
  this.postCommitCallbacks.length = 0;
  return callbacks;
}
```
or
```ts
async runPostCommitCallbacksAsync(): Promise<void> {
  const callbacks = this.postCommitCallbacks;
  this.postCommitCallbacks.length = 0;
  await Promise.all(callbacks.map(callback => callback()));
}
``` 
depending on whether the caller of the query context should be responsible for running the callbacks and if the query context should just be responsible for holding onto some state.

Reply:
Good suggestion! I'll modify it a bit to make all callback internals private to that class.